cdc/sink/dmlsink/mq/dmlproducer/pulsar_dml_producer.go

```diff
@@ -0,0 +1,312 @@
+			zap.String("changefeed", p.id.ID))
+		return
+	}
+	close(p.failpointCh)
+	// Notify the run loop to exit.
+	close(p.closedChan)
+	p.closed = true
+
+	start := time.Now()
+	keys := p.producers.Keys()
+	for _, topic := range keys {
+		p.producers.Remove(topic) // callback func will be called
+		topicName, _ := topic.(string)
+		log.Info("Async client closed in pulsar "+
+			"DML producer", zap.Duration("duration", time.Since(start)),
+			zap.String("namespace", p.id.Namespace),
+			zap.String("changefeed", p.id.ID), zap.String("topic", topicName))
+	}
+
+	p.client.Close()
+}
+
+// newProducer creates a pulsar producer
+// One topic is used by one producer
+func newProducer(
+	pConfig *config.PulsarConfig,
+	client pulsar.Client,
+	topicName string,
+) (pulsar.Producer, error) {
+	po := pulsar.ProducerOptions{
+		Topic: topicName,
+	}
+	if pConfig.BatchingMaxMessages != nil {
+		po.BatchingMaxMessages = *pConfig.BatchingMaxMessages
+	}
+	if pConfig.BatchingMaxPublishDelay != nil {
+		po.BatchingMaxPublishDelay = pConfig.BatchingMaxPublishDelay.Duration()
+	}
+	if pConfig.CompressionType != nil {
+		po.CompressionType = pConfig.CompressionType.Value()
+		po.CompressionLevel = pulsar.Default
+	}
+	if pConfig.SendTimeout != nil {
+		po.SendTimeout = pConfig.SendTimeout.Duration()
+	}
+
+	producer, err := client.CreateProducer(po)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("new pulsar producer ok", zap.String("topic:", topicName))
+
+	return producer, nil
+}
+
+func (p *pulsarDMLProducer) getProducer(topic string) (pulsar.Producer, bool) {
+	target, ok := p.producers.Get(topic)
+	if ok {
+		producer, ok := target.(pulsar.Producer)
+		if ok {
+			return producer, true
+		}
+	}
+	return nil, false
+}
+
+// GetProducerByTopic get producer by topicName
+func (p *pulsarDMLProducer) GetProducerByTopic(topicName string) (producer pulsar.Producer, err error) {
+	getProducer, ok := p.getProducer(topicName)
+	if ok && getProducer != nil {
+		return getProducer, nil
+	}
+
+	if !ok { // create a new producer for the topicName
+		producer, err = newProducer(p.pConfig, p.client, topicName)
+		if err != nil {
+			return nil, err
+		}
+		p.producers.Add(topicName, producer)
+	}
+
+	return producer, nil
+}
+
+// wrapperSchemaAndTopic wrapper schema and topic
+func (p *pulsarDMLProducer) wrapperSchemaAndTopic(m *common.Message) {
+	if m.Schema == nil {
+		if m.Protocol == config.ProtocolMaxwell {
+			mx := &maxwellMessage{}
+			err := json.Unmarshal(m.Value, mx)
+			if err != nil {
+				log.Error("unmarshal maxwell message failed", zap.Error(err))
+				return
+			}
+			if len(mx.Database) > 0 {
+				m.Schema = &mx.Database
+			}
+			if len(mx.Table) > 0 {
+				m.Table = &mx.Table
+			}
+		}
+		if m.Protocol == config.ProtocolCanal { // canal protocol set multi schemas in one topic
+			m.Schema = str2Pointer("multi_schema")
+		}
+	}
+}
+
+// maxwellMessage is the message format of maxwell
+type maxwellMessage struct {
```

Comment:
What is this ?

Reply:
The database name and table name are required in the monitoring report, but I can't get them from common.Message in version 6.5. I added this struct to parse and get these two values